### ISSUE TITLE
[>= 1.0.15] Cannot validate on file-like object

### ISSUE THREAD

**Issue by AxelVoitier** [sissaschool/xmlschema#145]
Hello,

I stumbled on an issue that apparently exists since 1.0.15.
In brief, when trying to use `xmlschema.validate()` with the XML doc being a file-like object (typically opened by `open`), then the parsing fails with:

> xml.etree.ElementTree.ParseError: no element found: line 1, column 0

Here is a test script (called `validate_test.py`) using example files provided by the projet (copied to current working dir):
```
import xmlschema


if __name__ == '__main__':
    f = open('vehicles.xml', 'r')
    xmlschema.validate(f, schema='vehicles.xsd')
```

Here is the output with 1.0.15 (from installation to run):
```
$ mkvirtualenv -p /usr/bin/python3 xmlschema-1.0.15
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/lexa/.virtualenvs/xmlschema-1.0.15/bin/python3
Also creating executable in /home/lexa/.virtualenvs/xmlschema-1.0.15/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.

(xmlschema-1.0.15) $ pip install xmlschema==1.0.15
Collecting xmlschema==1.0.15
  Using cached https://files.pythonhosted.org/packages/bb/51/b20ad8d56f85e5af85dfe1010e93185b738aaf400a64b8c0903e9cab3cfd/xmlschema-1.0.15-py2.py3-none-any.whl
Collecting elementpath~=1.3.0
  Using cached https://files.pythonhosted.org/packages/c3/f9/585359e5a09be6f0b32c1480e0f3117f513c333527f269050bc1130dc9a7/elementpath-1.3.1-py2.py3-none-any.whl
Installing collected packages: elementpath, xmlschema
Successfully installed elementpath-1.3.1 xmlschema-1.0.15

(xmlschema-1.0.15) $ python3 validate_test.py 
Traceback (most recent call last):
  File "validate_test.py", line 7, in <module>
    xmlschema.validate(f, schema='vehicles.xsd')
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/documents.py", line 74, in validate
    source, schema = get_context(xml_document, schema, cls, locations, base_url, defuse, timeout, lazy)
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/documents.py", line 40, in get_context
    source = XMLResource(source, defuse=defuse, timeout=timeout, lazy=lazy)
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/resources.py", line 253, in __init__
    self.source = source
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/resources.py", line 277, in __setattr__
    self._root, self._document, self._text, self._url = self._fromsource(value)
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/resources.py", line 335, in _fromsource
    document = self.parse(source)
  File "/home/lexa/.virtualenvs/xmlschema-1.0.15/lib/python3.6/site-packages/xmlschema/resources.py", line 436, in parse
    return ElementTree.parse(source)
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 1196, in parse
    tree.parse(source, parser)
  File "/usr/lib/python3.6/xml/etree/ElementTree.py", line 597, in parse
    self._root = parser._parse_whole(source)
xml.etree.ElementTree.ParseError: no element found: line 1, column 0

(xmlschema-1.0.15) $
```

And here is the same but with 1.0.14:
```
$ mkvirtualenv -p /usr/bin/python3 xmlschema-1.0.14
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/lexa/.virtualenvs/xmlschema-1.0.14/bin/python3
Also creating executable in /home/lexa/.virtualenvs/xmlschema-1.0.14/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.

(xmlschema-1.0.14) $ pip install xmlschema==1.0.14
Collecting xmlschema==1.0.14
  Using cached https://files.pythonhosted.org/packages/bc/07/17014ee90500f8028070e0697d17eb97851612afa48e8a98bd3254797c84/xmlschema-1.0.14-py2.py3-none-any.whl
Collecting elementpath~=1.2.0
  Using cached https://files.pythonhosted.org/packages/45/1b/dcd7204c046d4da94f8bb20f08788c26fb36815db1cf0b3b2c2b19ed2f25/elementpath-1.2.2-py2.py3-none-any.whl
Installing collected packages: elementpath, xmlschema
Successfully installed elementpath-1.2.2 xmlschema-1.0.14

(xmlschema-1.0.14) $ python3 validate_test.py 

(xmlschema-1.0.14) $ 
```

Cheers,
Axel

**Comment by brunato** [comment]
Hi,
this is due to a missing rewind of the file-object, that is used twice (to fetch locations and then to build the XML resource). Maybe the best solution is to move the following lines at beginning of the *get_context* function, just before the *try-except* statement:

https://github.com/sissaschool/xmlschema/blob/436567a8ad055445ed298c7a258e24511511698e/xmlschema/documents.py#L39-L40

In any case a fix will for this will be available into the next release.
Thank you
Davide

**Comment by brunato** [comment]
Fix available with release v1.0.17+.
Bye
